src/server/transaction.cc

```
@@ -847,90 +853,60 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
+    // 1. The phase is still active
+    // 2. We're still armed (i.e. nobody else run us on this shard)
+    // 3. We're still in the same phase as we expect (i.e. we didn't miss a full cycle)
 
-    bool is_armed = IsArmedInShard(shard->shard_id());
-    // First we check that this shard should run a callback by checking IsArmedInShard.
-    if (is_armed) {
-      uint32_t seq_after = seqlock_.load(memory_order_relaxed);
-
-      DVLOG(3) << "PollExecCb " << DebugId() << " sid(" << shard->shard_id() << ") "
-               << run_count_.load(memory_order_relaxed);
-
-      // We also make sure that for multi-operation transactions like Multi/Eval
-      // this callback runs on a correct operation. We want to avoid a situation
-      // where the first operation is executed and the second operation is armed and
-      // now this callback from the previous operation finally runs and calls PollExecution.
-      // It is usually ok, but for single shard operations we abuse index 0 in shard_data_
-      // Therefore we may end up with a situation where this old callback runs on shard 7,
-      // accessing shard_data_[0] that now represents shard 5 for the next operation.
-      // seqlock provides protection for that so each cb will only run on the operation it has
-      // been tasked with.
-      // We also must first check is_armed and only then seqlock. The first check ensures that
-      // the coordinator thread crossed
-      // "run_count_.store(unique_shard_cnt_, memory_order_release);" barrier and our seqlock_
-      // is valid.
-      if (seq_after == seq) {
-        // shard->PollExecution(this) does not necessarily execute this transaction.
-        // Therefore, everything that should be handled during the callback execution
-        // should go into RunInShard.
+    EngineShard* shard = EngineShard::tlocal();
+    if (IsArmedInShard(shard->shard_id())) {
+      if (run_barrier_.IsMyPhase(phase)) {
```

Actually, why do we need the seqlock at all if we check `IsArmedInShard()` also for direct ooo runs in PollExecution? We'll silently skip it and it won't run twice 🤔

---

you deleted the comment below that explains why we need to check the seqlock. I would like to restore it unless you think it's wrong